[sources-ui] STPSourceInfoViewController

bg-stripe opened the pull request:
re-r? @bdorfman-stripe 

My previous PR (https://github.com/stripe/stripe-ios/pull/555) got corrupted somehow after rebasing, so I had to make a new PR. It's the same one you reviewed already, only I've moved source-specific logic into subclasses of `STPSourceInfoDataSource`.

bdorfman-stripe commented on Stripe/PublicHeaders/STPSourceInfoViewController.h:
```diff
@@ -0,0 +1,59 @@
+/**
+ *  The view controller's delegate. This must be set before showing the view controller in order for it to work properly. @see STPSourceInfoViewControllerDelegate
+ */
+@property(nonatomic, weak) id<STPSourceInfoViewControllerDelegate> delegate;
```
I know we use delegates pretty much everywhere else, but what do you think about using completion blocks instead? Seems a bit easier for integrators to set up, and there are no callbacks here besides finished or cancelled.

bg-stripe replied:
Hm – I'd prefer to leave this as a delegate, just because it's the pattern we've been using elsewhere. I think if we decide to move to completion blocks we should do it across STPAddCardVC, STPAddSourceVC, and STPSourceInfoVC, since they all just have finished and cancelled callbacks.

bdorfman-stripe replied:
Ok. I think it might be worth doing that conversion at some point? But we can punt that for later.

bdorfman-stripe commented on Stripe/STPPickerTableViewCell.h:
```diff
@@ -0,0 +1,30 @@
+
+@protocol STPPickerDataSource <NSObject>
+
+- (NSInteger)numberOfRows;
```
Do we want to name these something more specific to picker data source to avoid potential future method name collisions?

bg-stripe replied:
good point – I'll update the names to things like`numberOfRowsInPicker`

bdorfman-stripe commented on Stripe/PublicHeaders/STPSourceInfoViewController.h:
```diff
@@ -0,0 +1,59 @@
+ *  @param sourceParams     The source parameters to collect additional info for. @see STPSourceParams
+  *  @param theme           The theme to use to inform the view controller's visual appearance. @see STPTheme
+ */
+- (nullable instancetype)initWithSourceParams:(STPSourceParams *)sourceParams
```
Do you want to use our existing "prefilled user information" class here? Or should those prefilled bits just be already set on the source params you are passing in?

bg-stripe replied:
Hm, I was imagining that we'd prefill the source params before passing them into STPSourceInfoVC, but it seems like it'd be better to accept `prefilledInformation` here – I'll update.

bdorfman-stripe replied:
Yeah, not sure what to do if the params have conflicting information to prefilledInformation? Would it be better to _just_ use prefilledInformation and return a new STPSourceParams? Or have a helper to fill source params from prefill?

bg-stripe replied:
I think it'd be best if you could immediately use the `STPSourceParams` object returned by this view controller, so it'd have to be prefilled with amount and returnURL, in addition to things in `prefilledInformation`. Maybe to avoid the conflicting information issue, we should just have an initializer that takes a source type, amount, returnURL, and prefilledInformation?

bdorfman-stripe replied:
Sure that makes sense.

Other VC's we provide init with payment configuration also right? Might be better to use that to pull the returnURL off?

bg-stripe replied:
ah yep, will use that instead of returnURL

bdorfman-stripe commented on Stripe/PublicHeaders/STPSourceInfoViewController.h:
```diff
@@ -0,0 +1,59 @@
+ *  Initializes a new `STPSourceInfoViewController` with the provided parameters. When the user submits the form, the view controller will return a copy of the provided `STPSourceParams` object to its delegate, updated with the information entered by the user.
+ *
+ *  @param sourceParams     The source parameters to collect additional info for. @see STPSourceParams
+  *  @param theme           The theme to use to inform the view controller's visual appearance. @see STPTheme
```
Should add a comment about returning nil if not a valid source type.

bg-stripe replied:
👍

bdorfman-stripe commented on Tests/Tests/STPSourceInfoViewControllerTest.m:
```diff
@@ -0,0 +1,109 @@
+
+@implementation STPSourceInfoViewControllerTest
+
+- (STPSourceInfoViewController *)sutWithParams:(STPSourceParams *)params {
```
What does `sut` stand for?

bg-stripe replied:
ah sorry, i use that jargon out of habit:
https://en.wikipedia.org/wiki/System_under_test

bdorfman-stripe commented on Tests/Tests/STPSourceInfoViewControllerTest.m:
```diff
@@ -0,0 +1,109 @@
+@property(nonatomic)STPSourceInfoDataSource *dataSource;
+@end
+
+@interface STPSourceInfoViewControllerTest : XCTestCase
```
Should this have snapshot tests? (can wait to add until designs are more final)

bg-stripe replied:
Yep – I was thinking we'd add snapshot tests when we do localization.

bdorfman-stripe commented:
Done with my initial review. Sorry for doing single comments instead of grouped github review.

bg-stripe commented:
@bdorfman-stripe addressed your comments in 3fe8572, and also added a new `completeSourceParams` property that you can check after initializing `STPSourceInfoVC` to determine if you need to present the view controller.

bdorfman-stripe commented on Stripe/STPBancontactSourceInfoDataSource.m:
```diff
@@ -27,7 +27,7 @@ - (instancetype)initWithSourceParams:(STPSourceParams *)sourceParams {
 }
 
-- (STPSourceParams *)completedSourceParams {
+- (STPSourceParams *)completeSourceParams {
```
Should we build this validation into STPSourceParams somehow? Or have a separate validator class as well? Instead of having it in this VC-specific data source type?

bg-stripe replied:
Hm, so I think of these methods as being pretty tied to the VC – they construct STPSourceParams from the data source's contents, and then perform some very light validation that just checks whether the fields in the form have been filled.

I'm not sure there's a need for a more in-depth STPSourceParams validator at this point – STPSourceInfoVC will do its best to ensure the returned SourceParams object is filled, and I think we can account for edge cases like not setting your returnURL by printing an error when you initialize the view controller.
